rfcs/proposed/task_group_dynamic_dependencies/implementation_details.md

```diff
@@ -0,0 +1,724 @@
+public:
+    task_tracker() : m_state(nullptr) {}
+
+    task_tracker(const task_tracker& other) : m_state(other.m_state) {
+        if (m_state) m_state->reserve();
+    }
+    task_tracker& operator=(const task_tracker& other) { /*same logic*/ }
+
+    task_tracker(task_tracker&& other) : m_state(other.m_state) {
+        other.m_state = nullptr;
+    }
+    task_tracker& operator=(task_tracker&& other) { /*same logic*/ }
+
+    task_tracker(const task_handle& th) : m_state(th ? th.get_dynamic_state() : nullptr) {
+        if (m_state) m_state->reserve();
+    }
+    task_tracker& operator=(const task_handle& th) { /*same logic*/ }
+
+    ~task_tracker() { if (m_state) m_state->release(); }
+};
+
+}
+}
+```
+
+### `task_dynamic_state` in details
+
+As it was mentioned above, `task_dynamic_state` class implements the tracking of task's state (completed or not completed),
+the list of successors and the post-transfer actions. It has the following layout:
+
+```cpp
+class task_dynamic_state {
+private:
+    task_with_dynamic_state*           m_task;
+    std::atomic<successors_list_node*> m_successors_list_head;
+    std::atomic<continuation_vertex*>  m_continuation_vertex;
+    std::atomic<task_dynamic_state*>   m_new_dynamic_state;
+    std::atomic<std::size_t>           m_num_references;
+    small_object_allocator             m_allocator;
+};
+```
+
+`m_task` is a pointer to the task, with which the current dynamic state is associated.
+
+`m_successors_list_head` is an atomic pointer to the head of the successor's list of the currently served task. It is also used
+as a marker of the task completion. It would be described in details in the following sections.
+
+`m_continuation_vertex` is an atomic pointer to the  `continuation_vertex` object associated with the currently served task. The purpose of the
+`continuation_vertex` is described in the following sections.
+
+`m_new_dynamic_state` is an atomic pointer to the other dynamic state instance. It is used when the currently served task calls
+`transfer_successors_to(new_task)` while executing. `m_new_dynamic_state` will point to `new_task`s dynamic state after doing the transfer.
+It is described in details in the section about transferring the successors.
+
+`m_num_references` and `m_allocator` manage the lifetime of the dynamic state.
+
+The stored reference counter is increased when:
+* `task_dynamic_state` associated with the task is created, the reference is reserved for the task object.
+* a `task_tracker` instance for the task is created (both non-empty tracker created from `task_handle` and copy of the existing one).
+* `transfer_successors_to(new_task)` is called, the dynamic state of the currently executing task reserves a reference on the
+  dynamic state of `new_task`. See [lifetime issue](#dynamic-state-lifetime-issue) section for more details.
+
+The stored reference counter is decreased when:
+* The associated task is completed, before destroying the task instance
```

Comment:
Just want to confirm that this is safe to decrement before the task is destroyed because once a task starts executing the only way to refer to a task_dynamic_state is through a task_tracker not a task.